R/compat-types-check.R

```diff
@@ -53,85 +113,181 @@ is_number <- function(x) {
+  if (is_call(x)) {
+    return(invisible(NULL))
+  }
+  if (allow_null && is_null(x)) {
+    return(invisible(NULL))
   }
+
+  stop_input_type(
+    x,
+    "a defused call",
+    ...,
+    allow_null = allow_null,
+    arg = arg,
+    call = call
+  )
 }
 
 check_environment <- function(x,
                               ...,
-                              what = "an environment",
+                              allow_null = FALSE,
                               arg = caller_arg(x),
                               call = caller_env()) {
-  if (!is_environment(x)) {
-    stop_input_type(x, what, ..., arg = arg, call = call)
+  if (is_environment(x)) {
+    return(invisible(NULL))
+  }
+  if (allow_null && is_null(x)) {
+    return(invisible(NULL))
   }
+
+  stop_input_type(
+    x,
+    "an environment",
+    ...,
+    allow_null = allow_null,
+    arg = arg,
+    call = call
+  )
 }
 
 check_function <- function(x,
                            ...,
-                           what = "a function",
+                           allow_null = FALSE,
                            arg = caller_arg(x),
                            call = caller_env()) {
-  if (!is_function(x)) {
-    stop_input_type(x, what, ..., arg = arg, call = call)
+  if (is_function(x)) {
+    return(invisible(NULL))
   }
+  if (allow_null && is_null(x)) {
+    return(invisible(NULL))
+  }
+
+  stop_input_type(
+    x,
+    "a function",
+    ...,
+    allow_null = allow_null,
+    arg = arg,
+    call = call
+  )
 }
 
 check_closure <- function(x,
-                           ...,
-                           what = "an R function",
-                           arg = caller_arg(x),
-                           call = caller_env()) {
-  if (!is_closure(x)) {
-    stop_input_type(x, what, ..., arg = arg, call = call)
+                          ...,
+                          allow_null = FALSE,
+                          arg = caller_arg(x),
+                          call = caller_env()) {
+  if (is_closure(x)) {
+    return(invisible(NULL))
+  }
+  if (allow_null && is_null(x)) {
+    return(invisible(NULL))
   }
+
+  stop_input_type(
+    x,
+    "an R function",
+    ...,
+    allow_null = allow_null,
+    arg = arg,
+    call = call
+  )
 }
 
 check_formula <- function(x,
                           ...,
-                          what = "a formula",
+                          allow_null = FALSE,
                           arg = caller_arg(x),
                           call = caller_env()) {
-  if (!is_formula(x)) {
-    stop_input_type(x, what, ..., arg = arg, call = call)
+  if (is_formula(x)) {
+    return(invisible(NULL))
+  }
+  if (allow_null && is_null(x)) {
+    return(invisible(NULL))
   }
+
+  stop_input_type(
+    x,
+    "a formula",
+    ...,
+    allow_null = allow_null,
+    arg = arg,
+    call = call
+  )
 }
 
 
 # Vectors -----------------------------------------------------------------
 
-# TODO: Restrict missing and special values
-
 check_character <- function(x,
```

Comment:
Should we remove this since it's the only function that checks for vectors?

Reply:
It's there because I've used it somewhere. I meant to treat this file as append-only storage for check functions we use in our packages. Does this make sense?

Reply:
Now that we're expanding the usage of this file, it might make sense to be a bit more deliberative about adding functions here. But if you think removing it (or moving to a different file if only rlang uses it), then we can keep it here.

Reply:
I think we could keep it there as a template and basis of discussion for vector checks.